examples/frost.c

```diff
@@ -0,0 +1,349 @@
+     * `secp256k1_xonly_pubkey_tweak_add` (after converting it to an xonly pubkey
+     * if necessary with `secp256k1_xonly_pubkey_from_pubkey`). */
+
+    /* Now we convert the output_pk to an xonly pubkey to allow to later verify
+     * the Schnorr signature against it. For this purpose we can ignore the
+     * `pk_parity` output argument; we would need it if we would have to open
+     * the taproot commitment. */
+    if (!secp256k1_xonly_pubkey_from_pubkey(ctx, agg_pk, NULL, &output_pk)) {
+        return 0;
+    }
+    return 1;
+}
+
+/* Sign a message hash with the given threshold and aggregate shares and store
+ * the result in sig */
+int sign(const secp256k1_context* ctx, struct signer_secrets *signer_secrets, struct signer *signer, const unsigned char* msg32, secp256k1_xonly_pubkey *agg_pk, unsigned char *sig64, const secp256k1_frost_tweak_cache *cache) {
+    int i;
+    int signer_id = 0;
+    int signers[THRESHOLD];
+    int is_signer[N_SIGNERS];
+    const secp256k1_frost_pubnonce *pubnonces[THRESHOLD];
+    const secp256k1_xonly_pubkey *pubkeys[THRESHOLD];
+    const secp256k1_frost_partial_sig *partial_sigs[THRESHOLD];
+    unsigned int seed;
+
+    for (i = 0; i < N_SIGNERS; i++) {
+        FILE *frand;
+        unsigned char seckey[32];
+        unsigned char session_id[32];
+        /* Create random session ID. It is absolutely necessary that the session ID
+         * is unique for every call of secp256k1_frost_nonce_gen. Otherwise
+         * it's trivial for an attacker to extract the secret key! */
+        frand = fopen("/dev/urandom", "r");
+        if(frand == NULL) {
+            return 0;
+        }
+        if (!fread(session_id, 32, 1, frand)) {
+            fclose(frand);
+            return 0;
+        }
+        fclose(frand);
+        if (!secp256k1_keypair_sec(ctx, seckey, &signer_secrets[i].keypair)) {
+            return 0;
+        }
+        /* Initialize session and create secret nonce for signing and public
+         * nonce to send to the other signers. */
+        if (!secp256k1_frost_nonce_gen(ctx, &signer_secrets[i].secnonce, &signer[i].pubnonce, session_id, &signer_secrets[i].agg_share, msg32, agg_pk, NULL)) {
+            return 0;
+        }
+        is_signer[i] = 0; /* Initialize is_signer */
+    }
+    /* Select a random subset of signers */
+    for (i = 0; i < THRESHOLD; i++) {
+        while (1) {
+            if (!fill_random((unsigned char*)&seed, sizeof(seed))) {
+                return 0;
+            }
+            signer_id = seed % N_SIGNERS;
+            /* Check if signer has already been assigned */
+            if (!is_signer[signer_id]) {
+                is_signer[signer_id] = 1;
+                signers[i] = signer_id;
+                break;
+            }
+        }
+        /* Mark signer as assigned */
+        pubnonces[i] = &signer[signer_id].pubnonce;
+        pubkeys[i] = &signer[signer_id].pubkey;
+    }
+    /* Signing communication round 1: Exchange nonces */
```

Comment:
If there is only signing round in this variant, the reference to the 1st can be dropped, unless there are multiples other rounds.

Reply:
The first signing communication round is the nonce exchange, and the second round is the partial signature exchange [below](https://github.com/ElementsProject/secp256k1-zkp/pull/138/files/fd703109a76dcb88cf35b8e602f54ade2a8bb9ef#diff-2d1c5749d9c83eb8473d43bed1b741f624529dd944ce21fc52614ac7e4bd4921R265-R266). 

There's an inconsistency in the labeling that I will fix: `Signing communication round` and `Communication round`.